graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+    // select the
+    // most restrictive cache policy from those types.
+
+    if (unwrappedReturnType.getName().equals(_Entity.typeName)) {
+      Object representations = parameters.getExecutionStepInfo().getArgument(_Entity.argumentName);
+
+      if (representations instanceof List) {
+        typesFromEntitiesArgument(
+                representations, parameters.getExecutionContext().getGraphQLSchema())
+            .stream()
+            .map(type -> hintCache.getHint((GraphQLCompositeType) type))
+            .filter(Optional::isPresent)
+            .forEach(hint -> fieldPolicy.restrict(hint.get()));
+      }
+    }
+
+    // Cache hint on the return type of this field if it's a composite type
+
+    if (unwrappedReturnType instanceof GraphQLCompositeType) {
+      Optional<Hint> hint = hintCache.getHint((GraphQLCompositeType) unwrappedReturnType);
+
+      if (hint.isPresent()) {
+        fieldPolicy.replace(hint.get());
+        inheritMaxAge = hint.get().getInheritMaxAge().orElse(false);
+      }
+    }
+
+    // Cache hint on the field itself
+
+    Optional<Hint> fieldHint = hintCache.getHint(parameters.getField());
+    if (fieldHint.isPresent()) {
+      Hint hint = fieldHint.get();
+
+      // If inheritMaxAge is true, take note of that to avoid setting the default max age in the
+      // next step.
+      // This does allow setting the cache scope though.
+      //
+      // Note: setting both maxAge and inheritMaxAge doesn't make sense.
+      if (hint.getInheritMaxAge().orElse(false) && !fieldPolicy.hasMaxAge()) {
+        inheritMaxAge = true;
+        fieldPolicy.replace(hint.getScope());
+      } else {
+        fieldPolicy.replace(hint);
+      }
+    }
+
+    // If this field returns a composite type or is a root field and
+    // we haven't seen an explicit maxAge hint, set the maxAge to 0
+    // (uncached) or the default if specified in the constructor.
+    // (Non-object fields by default are assumed to inherit their
+    // cacheability from their parents. But on the other hand, while
+    // root non-object fields can get explicit hints from their
+    // definition on the Query/Mutation object, if that doesn't exist
+    // then there's no parent field that would assign the default
+    // maxAge, so we do it here.)
+    //
+    // You can disable this on a non-root field by writing
+    // `@cacheControl(inheritMaxAge: true)` on it. If you do this,
+    // then its children will be treated like root paths, since there
+    // is no parent maxAge to inherit.
+
+    if (!fieldPolicy.hasMaxAge()
+        && ((unwrappedReturnType instanceof GraphQLCompositeType && !inheritMaxAge)
+            || parameters.getExecutionStepInfo().getPath().isRootPath())) {
+      fieldPolicy.restrict(defaultMaxAge);
+    }
+
+    return new SimpleInstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onCompleted(ExecutionResult result, Throwable t) {
+        state.restrictCachePolicy(fieldPolicy);
```

Comment:
If you're not implementing dynamic cache control, then you don't actually need to do this on the onCompleted; you can just do it at the end of beginField.

That said, I don't think it would be too hard to implement dynamic cache control! I'd suggest sticking something on the GraphQLContext.  Then you could have a static method `CacheControlInstrumentation.cacheControlContext(GraphQLContext c) { return (CacheControlContext) c.get(CACHE_CONTROL_CONTEXT) }` to let resolvers access it.

You could then follow the pattern of the TS implementation where the _Entity code is just using dynamic cache control instead of being special-cased.

Reply:
this is good — i'll try implementing it now that i'm using the context for things (instead of extensions).